### PR TITLE
VITIS-3524 - AIE Partition support for xclbin containers and xclbinutil

### PR DESCRIPTION
**Pull request by rozumx**
#### Problem solved by the commit
Enhanced xclbin containers and xclbinutil to support AIE Partition metadata.

Work Done
------------
+ Refactored readXclBinBinar() and readPayload() to use std::istream classe instead of std::fstream
+ Created new section AIE_PARTITION and its supporting functions
+ Updated flow to auto create AIE_PARTITION section if AIT_METADATA contains AIE Partition information
+ Update XclBinUtilities::TRACE to accept boost::format types
+ Added --skip-aie-partition-insertion option to skip automatically added the AIE_PARTITION section if one isn't found
+ Created unit-tests to add / dump AIE_PARTITIONS
+ Created unit-tests to automatically extract AIE_PARTITION metadata from AIE_METADATA

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
N/A

#### How problem was solved, alternative solutions (if any) and why they were rejected
N/A

#### Risks (if any) associated the changes in the commit
none - incremental addition

#### What has been tested and how, request additional testing if necessary
+ Several unit-tests have been added for this new functionality
+ Manual testing

#### Documentation impact (if any)
N/A

### CODE REVIEW

File: src/runtime_src/tools/xclbinutil/SectionAIEPartition.cxx [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  _buf.write(sStringBlock.c_str(), sStringBlock.size());
+}
+
+bool
+SectionAIEPartition::doesSupportAddFormatType(FormatType _eFormatType) const
+{
+  if (_eFormatType == FT_JSON) {
+    return true;
+  }
+  return false;
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/SectionAIEPartition.cxx:159]:
```suggestion
  return (_eFormatType == FT_JSON);
```

File: src/runtime_src/tools/xclbinutil/SectionAIEPartition.cxx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+bool
+SectionAIEPartition::doesSupportDumpFormatType(FormatType _eFormatType) const
+{
+  if ((_eFormatType == FT_JSON) ||
+      (_eFormatType == FT_HTML) ||
+      (_eFormatType == FT_RAW)) {
+    return true;
+  }
+
+  return false;
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/SectionAIEPartition.cxx:171]:
```suggestion
  return (_eFormatType == FT_JSON) ||
      (_eFormatType == FT_HTML) ||
      (_eFormatType == FT_RAW);
```

File: src/runtime_src/tools/xclbinutil/SectionAIEPartition.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// -------- C L A S S :   S e c t i o n A I E P a r t i t i o n --------------
+
+class SectionAIEPartition : public Section {
+ public:
+  SectionAIEPartition();
+  virtual ~SectionAIEPartition();
+
+ public:
+  virtual bool doesSupportAddFormatType(FormatType _eFormatType) const;
+  virtual bool doesSupportDumpFormatType(FormatType _eFormatType) const;
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/SectionAIEPartition.h:35]:
```suggestion
  bool doesSupportAddFormatType(FormatType _eFormatType) const override;
  bool doesSupportDumpFormatType(FormatType _eFormatType) const override;
```
The compiler actually uses override to validate that you indeed are overriding some exact match

File: src/runtime_src/tools/xclbinutil/SectionAIEPartition.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+ protected:
+  virtual void marshalToJSON(char* _pDataSection, unsigned int _sectionSize, boost::property_tree::ptree& _ptree) const;
+  virtual void marshalFromJSON(const boost::property_tree::ptree& _ptSection, std::ostringstream& _buf) const;
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/SectionAIEPartition.h:39]:
```suggestion
  void marshalToJSON(char* _pDataSection, unsigned int _sectionSize, boost::property_tree::ptree& _ptree) const overide;
  void marshalFromJSON(const boost::property_tree::ptree& _ptSection, std::ostringstream& _buf) const override;
```

File: src/runtime_src/tools/xclbinutil/SectionSoftKernel.h [hunk 5]
```diff
@@ -44,7 +44,7 @@ class SectionSoftKernel : public Section {
   virtual bool doesSupportAddFormatType(FormatType _eFormatType) const;
   virtual bool supportsSubSection(const std::string &_sSubSectionName) const;
   virtual bool subSectionExists(const std::string &_sSubSectionName) const;
-  virtual void readXclBinBinary(std::fstream& _istream, const struct axlf_section_header& _sectionHeader);
+  virtual void readXclBinBinary(std::istream& _istream, const struct axlf_section_header& _sectionHeader);
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/SectionSoftKernel.h:47]: override ?

File: src/runtime_src/tools/xclbinutil/SectionVenderMetadata.h [hunk 6]
```diff
@@ -32,7 +32,7 @@ class SectionVenderMetadata : public Section {
   virtual bool doesSupportAddFormatType(FormatType _eFormatType) const;
   virtual bool supportsSubSection(const std::string& _sSubSectionName) const;
   virtual bool subSectionExists(const std::string& _sSubSectionName) const;
-  virtual void readXclBinBinary(std::fstream& _istream, const struct axlf_section_header& _sectionHeader);
+  virtual void readXclBinBinary(std::istream& _istream, const struct axlf_section_header& _sectionHeader);
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/SectionVenderMetadata.h:35]: override ?

File: src/runtime_src/tools/xclbinutil/XclBinUtilities.cxx [hunk 7]
```diff
@@ -1034,6 +1039,76 @@ XclBinUtilities::createMemoryBankGrouping(XclBin & xclbin)
...
 
 
+void 
+XclBinUtilities::createAIEPartition(XclBin & xclbin)
+{
+  const boost::property_tree::ptree ptEmpty;
+
+  // -- DRC checks
+  const auto pAIEMetadataSection = xclbin.findSection(AIE_METADATA);
+  if (pAIEMetadataSection == nullptr)
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/XclBinUtilities.cxx:1049]:
```suggestion
  if (!pAIEMetadataSection)
```

**Review by larry9523 — APPROVED** [review]

**Review by AShivangi — APPROVED** [review]

File: src/runtime_src/tools/xclbinutil/SectionAIEPartition.cxx (hunk 1 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/SectionAIEPartition.cxx]: Changes made

File: src/runtime_src/tools/xclbinutil/SectionAIEPartition.cxx (hunk 2 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/SectionAIEPartition.cxx]: Changes made

File: src/runtime_src/tools/xclbinutil/SectionVenderMetadata.h (hunk 6 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/SectionVenderMetadata.h]: Changes made

File: src/runtime_src/tools/xclbinutil/XclBinUtilities.cxx (hunk 7 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/XclBinUtilities.cxx]: Changes made

**Pull request merged**